test/cpos/test_cpo_connect.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) Lucian Radu Teodorescu
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <catch2/catch.hpp>
+#include <execution.hpp>
+#include "test_common/receivers.hpp"
+
+namespace ex = std::execution;
+
+template <typename R>
+struct op_state {
+  int val_;
+  R recv_;
+
+  friend void tag_invoke(ex::start_t, op_state& self) noexcept {
+    ex::set_value((R &&) self.recv_, self.val_);
+  }
+};
+
+struct my_sender {
+  template <template <class...> class Tuple, template <class...> class Variant>
+  using value_types = Variant<Tuple<>>;
+  template <template <class...> class Variant>
+  using error_types = Variant<std::exception_ptr>;
+  static constexpr bool sends_done = false;
+
+  int value_{0};
+
+  template <typename R>
+  friend op_state<R> tag_invoke(ex::connect_t, my_sender&& s, R&& r) {
+    return {s.value_, (R &&) r};
+  }
+};
+
+TEST_CASE("can call connect on an appropriate types", "[cpo][cpo_connect]") {
+  auto op = ex::connect(my_sender{10}, expect_value_receiver{10});
+  ex::start(op);
+  // the receiver will check the received value
+}
+
+TEST_CASE("cannot connect sender with invalid receiver", "[cpo][cpo_connect]") {
+  // REQUIRE_FALSE(std::tag_invocable<ex::connect_t, my_sender, int>);
+  // TODO: this should not work
+  // invalid check:
+  REQUIRE(std::tag_invocable<ex::connect_t, my_sender, int>);
```

Comment:
This works because you have not constrained `my_sender::tag_invoke(connect_t...)` to accept only receivers. It is not a problem with the `connect` CPO or with the `tag_invoke` machinery.

Reply:
Please see [9.6.2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2300r3.html#spec-execution.senders.connect), bullet 2:
>  If `R` does not satisfy `execution::receiver`, `execution::connect(s, r)` is ill-formed. 

I agree that for the vast majority of users this is not a problem; I was just trying to test this edge-case

Reply:
Using connect as an invokable and using it as a tag in a call to tag_invoke are two different things.

Reply:
Sorry to be late.

I had the exact same issue you are facing when I wrote my initial tests. So I would like to comment on why the code in question is correct, but the tests are actually invalid.

In a nutshell you are getting the right answer to the wrong question.

```std::tag_invocable<ex::connect_t, my_sender, int>``` answers the question of whether the user declared an overload of `tag_invoke(ex::connect_t, mysender, int)`

It does however *not* answer the question whether that is a valid invocation of `ex::connect` because the constraints are checked in the CPO and not in the user provided implementation.

The *right* check it to test `std::invocable<ex::connect_t, my_sender, int>` as that will check whether we can actually call `ex::connect_t(my_sender, int)` and applies all the constraint checks.

From my experience writing some of the tests we should replace most of the instances of `tag_invocable<CPO, ...>` with `invocable<CPO, ...>` except when we want to make clear, that the CPO has additional constraints as in this example, where we should expand to: 

```cpp
  REQUIRE(std::tag_invocable<ex::connect_t, my_sender, int>);
  REQUIRE(!std::invocable<ex::connect_t, my_sender, int>);
```

Reply:
Thank you for the explanation. Makes a lot of sense.
It was also bothering me that I didn't properly test the connect CPO, but I didn't think of using std::invocable.

I will make this change in a followup patch.

Reply:
Stupid me. If I change this to `std::invocable` the test works as expected. I could have sworn that `connect` doesn't constrain the receiver to model the `receiver` concept. But, looking at the git history it's just my imagination.

Sorry, and thank you again!

Reply:
See #372 for the fix